mxcubecore/model/protocols/epics.py

```diff
@@ -0,0 +1,47 @@
+"""Models the `epics` section of YAML hardware configuration file.
+
+Provides an API to read configured EPICS channels.
+"""
+from typing import Optional, Iterable, Tuple, Dict
+from pydantic import BaseModel
+
+
+class Channel(BaseModel):
+    """EPICS channel configuration."""
+
+    suffix: Optional[str]
+    poll: Optional[int]
+
+
+class Prefix(BaseModel):
+    """Configuration of an EPICS prefix section."""
+
+    channels: Optional[Dict[str, Optional[Channel]]]
+
+    def get_channels(self) -> Iterable[Tuple[str, Channel]]:
```

Comment:
I think it should be `Iterator` and not `Iterable`.

```suggestion
    def get_channels(self) -> Iterator[Tuple[str, Channel]]:
```

Reply:
It is valid, generators _are_ Iterables:

```
>>> from typing import Iterable
>>> def foo():
...   yield 1
...   yield 3
>>> isinstance(foo(), Iterable)
True
```

Here I want to specify a more generic return type. It is _some_ kind of iterable. Today it's generator, but in the future perhaps it is a list.

Reply:
I see, I guess it is fine... I do not know for sure.

---

The rules I had read some time ago were like this:

* When you produce you should use as precise a type you can.
* When you consume you should use as generic a type as you can that still fulfils your usage.

so that is what I always follow and I never had any issue.

So for example, I would do something like this:

```python
def producer() -> list[int]:
    return [1, 2, 3]

def consumer(things: typing.Iterable[int]) -> None
    for thing in things:
        print(thing)

some_things = producer()
consumer(some_things)
```

If later the `producer` function is refactored into a generator (or iterator or any other iterable), I still do not need to change anything in `consumer`.

Reply:
A question purely for my education (you guys have studied more than I have). Reading you example one might be tempted to think that producer promises to return a list. Would it be OK to write code like 

list2 = list1 + producer() 

Or should one  instead write

list2 = list1 + list(producer())

in case producer was later refactored to return something else? Or is this principle only valid for certain contexts?

(I know you could write something like 

list2 = []; list2.extend(list1); list2.extend(producer()) 

but the question is whether you need to assume that the return type of producer() might be(come) any iterable at some later date.

Reply:
If you have

```
def producer() -> list[int]:
    return [1, 2, 3]
```

Then, it explicitly promises to return a list. Thus you can use return value as a list.

However, if you have:

```
def producer() -> Iterable[int]:
    return [1, 2, 3]
```

Then it does not promise a list. This can any type of iterable value. Then you should explicitly convert this to a list, if you a doing list operations on it.

Here is documentation on Iterable type annotation: https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable


However, bear in mind that none of this is enforced during runtime. Nothing in python stops you from doing this:

```
def producer() -> Iterable[int]:
    return False
```

Python will happily execute this and return a boolean to you.

The only way to find you that your code is not following the type annotations is to run a static checker tool, such as mypy.

Reply:
@elmjag Thanks, that makes sense. I know that Python does not respect type annotations, but I want to be a good boy and follow our rules, so I was trying to make sure I understood what our rules were.

Reply:
I suggest that we try out pyright: https://github.com/microsoft/pyright

Reply:
> I suggest that we try out pyright: https://github.com/microsoft/pyright

Yes, I agree. I wrote a new issue: https://github.com/mxcube/mxcubecore/issues/1053